frontend/__tests__/components/environment.spec.tsx

```diff
@@ -0,0 +1,62 @@
+import * as React from 'react';
+import { shallow, mount } from 'enzyme';
+
+import { EnvironmentPage } from '../../public/components/environment';
+
+describe('EnvironmentPage', () => {
+
+  const objects={
+    kind: '',
+    apiVersion: '',
+    metadata: {},
+    spec: {},
+    status: {}
+  };
+
+  const environmentPageRO=<EnvironmentPage
+    obj={objects}
+    rawEnvData={[ { 'env': [ { 'name': 'test', 'value': ':0', 'ID': 0 } ] } ]}
+    envPath={[]}
+    readOnly={true}
+  />;
+
+  const environmentPage=<EnvironmentPage
+    obj={objects}
+    rawEnvData={[ { 'env': [ { 'name': 'test', 'value': ':0', 'ID': 0 } ] } ]}
+    envPath={['test']}
+    readOnly={false}
+  />;
+
+  it('does not show help text when page read-only', () => {
+    const wrapper=shallow(environmentPageRO);
+    expect(wrapper.find('p').exists()).toEqual(false);
+  });
+
+  it('renders save and reload buttons when page not read-only', () => {
+    const wrapper=shallow(environmentPage);
+    expect(wrapper.find('.environment-buttons button').exists()).toEqual(true);
+  });
+
+  it('renders error message when error in state', () => {
+    const wrapper=shallow(environmentPage);
+    wrapper.setState({errorMessage: 'errorMessage'});
+    expect(wrapper.find('.environment-buttons p').text()).toContain('errorMessage');
+  });
+
+  it('renders error message when data is stale', () => {
+    const wrapper=shallow(environmentPage);
+    wrapper.setState({stale: 'stale'});
+    expect(wrapper.find('.environment-buttons p').text()).toContain('The information on this page is no longer current. Click Reload to update and lose edits, or Save Changes to overwrite.');
+  });
+
+  it('renders success message when data is updated successfully', () => {
```

Comment:
Positive/Negative on this one could be something like:

`renders success message when data is updated successfully`
`renders error message when data is not updated successfully`